--- lf_bak.c	2026-09-07 08:39:41.379690502 -0500
+++ lf.c	2026-09-07 17:12:09.492938144 -0500
@@ -1,8 +1,5 @@
-/** NOTICE: This file is part of the lf project, which is currently under
- * development. There are two reasons not to use it at this time. First,
- * it is not yet fully functional and may contain bugs or incomplete,
- * unoptimized code. Second, the API and features are still being finalized,
- * so using it now may lead to compatibility issues in the future as changes
- * are made. Once the project is more mature and stable, this file will be
- * ready for use. In the meantime, it serves as a work in progress and may
- * be subject to significant changes as development continues.
+/** ANNOUNCEMENT: This file is part of the lf project, which is currently being
+ * tested in anticipation of public release. The test suite consists of a test
+ * script, lf_tests.sh, which uses diff to compare the output with find. There
+ * is also an accompanying markdown file, lf_tests.md, that outlines the testing
+ * methodology, cases, and expected results.
@@ -10 +7,10 @@
- * Thank you for your patience and understanding.
+ * One test run resulted in six files out of 600,000 listed by find that weren't
+ * listed by lf. These were temporary files created by Google browser, Microsoft
+ * Edge, and Thunderbird. lf rejected them because their inodes were fictitious.
+ * find listed them without distinguising them from the other 599,993 normal
+ * files. With find, you would never know about them. As a design choice, lf
+ * segregates those files as errors. We aren't necessarily locked into our
+ * design choices. Your feedback and suggestions are always welcome.
+ *
+ * Feel free to run the test script on your own system, and if you encounter any
+ * issues, please report them to the author.
@@ -70 +75,0 @@
-// Search Filters
@@ -71,0 +77,24 @@
+    dev_t dev;
+    ino_t ino;
+} History;
+typedef struct TaskNode TaskNode;
+struct TaskNode {
+    History *history;    /**< Array of dev/ino pairs for cycle detection */
+    TaskNode *next_task; /**< Pointer to the next node in the queue */
+    int depth;           /**< Current depth in the directory tree */
+    char *dir_path;      /**< Directory path to process */
+}; /**< Queue TaskNode (for work-stealing) */
+
+typedef struct {
+    TaskNode *qhead;
+    TaskNode *qtail;
+    pthread_mutex_t queue_mutex;
+    pthread_cond_t cond_var;
+    pthread_mutex_t output_mutex;
+    pthread_t *threads;
+    atomic_int active_tasks;
+    int shut_down;
+    atomic_size_t file_count;
+    atomic_size_t error_count;
+    TerminationStatus termination_status;
+    unsigned int nthreads;
@@ -105 +134 @@
-} SearchFilters;
+} LfContext;
@@ -111,12 +139,0 @@
-typedef struct {
-    dev_t dev;
-    ino_t ino;
-} History;
-
-typedef struct TaskNode TaskNode;
-struct TaskNode {
-    History *history;    /**< Array of dev/ino pairs for cycle detection */
-    TaskNode *next_task; /**< Pointer to the next node in the queue */
-    int depth;           /**< Current depth in the directory tree */
-    char *dir_path;      /**< Directory path to process */
-}; /**< Queue TaskNode (for work-stealing) */
@@ -137,26 +153,0 @@
-typedef struct {
-    TaskNode *qhead;
-    TaskNode *qtail;
-    pthread_mutex_t queue_mutex;
-    pthread_cond_t cond_var;
-    pthread_mutex_t output_mutex;
-    atomic_int active_tasks;
-    int shut_down;
-    atomic_size_t file_count;
-    atomic_size_t error_count;
-    TerminationStatus termination_status;
-    unsigned int nthreads;
-} LfContext;
-
-TaskNode *qhead = NULL;
-TaskNode *qtail = NULL;
-pthread_mutex_t queue_mutex = PTHREAD_MUTEX_INITIALIZER;
-pthread_cond_t cond_var = PTHREAD_COND_INITIALIZER;
-pthread_mutex_t output_mutex = PTHREAD_MUTEX_INITIALIZER;
-atomic_int active_tasks = 0;
-int shut_down = 0;
-atomic_size_t file_count = 0;
-atomic_size_t error_count = 0;
-TerminationStatus termination_status = TS_ERROR;
-unsigned int nthreads;
-
@@ -169,5 +160,6 @@
-void debug_out(SearchFilters *, int, char **, int);
-bool init_find(SearchFilters *, int, char **);
-void sort_lf_output(SearchFilters *, int, char **);
-void enqueue_dir(TaskNode *);
-TaskNode *dequeue_dir();
+void debug_out(LfContext *lf, int, char **);
+
+bool init_find(LfContext *lf, int, char **);
+void sort_lf_output(LfContext *lf, int, char **);
+void enqueue_dir(LfContext *lf, TaskNode *);
+TaskNode *dequeue_dir(LfContext *lf);
@@ -175 +167 @@
-int scan_file(char *, const SearchFilters *, const unsigned char);
+int scan_file(char *, LfContext *lf, const unsigned char);
@@ -210 +202 @@
-    SearchFilters *f = state->input;
+    LfContext *lf = state->input;
@@ -216,2 +208,2 @@
-        parse_local_timestamp(arg, &f->after);
-        if (f->after && f->before && f->before < f->after) {
+        parse_local_timestamp(arg, &lf->after);
+        if (lf->after && lf->before && lf->before < lf->after) {
@@ -219 +211 @@
-            f->after = 0;
+            lf->after = 0;
@@ -223,2 +215,2 @@
-        parse_local_timestamp(arg, &f->before);
-        if (f->after && f->before && f->before < f->after) {
+        parse_local_timestamp(arg, &lf->before);
+        if (lf->after && lf->before && lf->before < lf->after) {
@@ -226 +218 @@
-            f->before = 0;
+            lf->before = 0;
@@ -230 +222 @@
-        f->count = true;
+        lf->count = true;
@@ -233 +225 @@
-                f->count_silently |= true;
+                lf->count_silently |= true;
@@ -238 +230 @@
-            f->max_depth = a_toi(arg, &a_toi_error);
+            lf->max_depth = a_toi(arg, &a_toi_error);
@@ -241 +233 @@
-        f->debug = true;
+        lf->debug = true;
@@ -248 +240 @@
-                    f->report_config = true;
+                    lf->report_config = true;
@@ -251 +243 @@
-                    f->report_info = true;
+                    lf->report_info = true;
@@ -254 +246 @@
-                    f->report_warnings = true;
+                    lf->report_warnings = true;
@@ -257 +249 @@
-                    f->report_errors = true;
+                    lf->report_errors = true;
@@ -260 +252 @@
-                    f->report_badlinks = true;
+                    lf->report_badlinks = true;
@@ -263 +255 @@
-                    f->report_trace = true;
+                    lf->report_trace = true;
@@ -266,6 +258,6 @@
-                    f->report_config = true;
-                    f->report_info = true;
-                    f->report_warnings = true;
-                    f->report_errors = true;
-                    f->report_badlinks = true;
-                    f->report_all = true;
+                    lf->report_config = true;
+                    lf->report_info = true;
+                    lf->report_warnings = true;
+                    lf->report_errors = true;
+                    lf->report_badlinks = true;
+                    lf->report_all = true;
@@ -274 +266 @@
-                    f->only_errors = true;
+                    lf->only_errors = true;
@@ -285,2 +277,2 @@
-        f->ere = strdup(arg);
-        f->flags |= LF_EXC_REGEX;
+        lf->ere = strdup(arg);
+        lf->flags |= LF_EXC_REGEX;
@@ -289 +281 @@
-        f->include_hidden = true; // Include hidden files
+        lf->include_hidden = true; // Include hidden files
@@ -291,2 +283,2 @@
-            f->hidden_only = true;
-            f->include_hidden = true;
+            lf->hidden_only = true;
+            lf->include_hidden = true;
@@ -294 +286 @@
-        f->flags &= ~(LF_HIDE); // Turn hide flag off
+        lf->flags &= ~(LF_HIDE); // Turn hide flag off
@@ -299 +291 @@
-        f->flags |= LF_ICASE;
+        lf->ignore_case = true;
@@ -302 +294 @@
-        f->follow_links = true; // follow symbolic links
+        lf->follow_links = true; // follow symbolic links
@@ -309 +301 @@
-                f->include_perms |= LF_ISGID;
+                lf->include_perms |= LF_ISGID;
@@ -312 +304 @@
-                f->include_perms |= LF_IRUSR;
+                lf->include_perms |= LF_IRUSR;
@@ -315 +307 @@
-                f->include_perms |= LF_ISUID;
+                lf->include_perms |= LF_ISUID;
@@ -318 +310 @@
-                f->include_perms |= LF_IWUSR;
+                lf->include_perms |= LF_IWUSR;
@@ -321 +313 @@
-                f->include_perms |= LF_IXUSR;
+                lf->include_perms |= LF_IXUSR;
@@ -329 +321 @@
-        f->sort_reverse = true;
+        lf->sort_reverse = true;
@@ -332,2 +324,2 @@
-        f->re = strdup(arg);
-        f->flags |= LF_REGEX;
+        lf->re = strdup(arg);
+        lf->flags |= LF_REGEX;
@@ -336 +328 @@
-        f->sort = true;
+        lf->sort = true;
@@ -339 +331 @@
-        f->file_size_min = (intmax_t)(a_to_ul(arg));
+        lf->file_size_min = (intmax_t)(a_to_ul(arg));
@@ -343 +335 @@
-            nthreads = a_toi(arg, &a_toi_error);
+            lf->nthreads = a_toi(arg, &a_toi_error);
@@ -350 +342 @@
-                f->include_types |= LF_BLK;
+                lf->include_types |= LF_BLK;
@@ -353 +345 @@
-                f->include_types |= LF_CHR;
+                lf->include_types |= LF_CHR;
@@ -356 +348 @@
-                f->include_types |= LF_DIR;
+                lf->include_types |= LF_DIR;
@@ -359 +351 @@
-                f->include_types |= LF_FIFO;
+                lf->include_types |= LF_FIFO;
@@ -362 +354 @@
-                f->include_types |= LF_LNK;
+                lf->include_types |= LF_LNK;
@@ -367 +359 @@
-                f->include_types |= LF_REG;
+                lf->include_types |= LF_REG;
@@ -370 +362 @@
-                f->include_types |= LF_SOCK;
+                lf->include_types |= LF_SOCK;
@@ -373 +365 @@
-                f->include_types |= LF_UNKNOWN;
+                lf->include_types |= LF_UNKNOWN;
@@ -381 +373 @@
-        f->user_name = strdup(arg);
+        lf->user_name = strdup(arg);
@@ -384,2 +376,2 @@
-            f->user_id = (uintmax_t)pwd->pw_uid;
-            f->flags |= LF_USER;
+            lf->user_id = (uintmax_t)pwd->pw_uid;
+            lf->flags |= LF_USER;
@@ -388 +380 @@
-            exit(termination_status);
+            exit(EXIT_FAILURE);
@@ -393 +385 @@
-        f->exec = strdup(arg);
+        lf->exec = strdup(arg);
@@ -414,0 +407,2 @@
+    int rc;
+    LfContext *lf = (LfContext *)calloc(1, sizeof(LfContext));
@@ -416,6 +410,21 @@
-    SearchFilters *f = (SearchFilters *)calloc(1, sizeof(SearchFilters));
-    f->ignore_case = false;
-    f->sort = false;
-    f->sort_reverse = false;
-    f->include_hidden = false; // By default, hidden files are suppressed. Use
-                               // -H to include them.
+    lf->qhead = NULL;
+    lf->qtail = NULL;
+    rc = pthread_mutex_init(&lf->queue_mutex, NULL);
+    if (rc != 0)
+        perror("Mutex initialization failed");
+    rc = pthread_cond_init(&lf->cond_var, NULL);
+    if (rc != 0)
+        perror("Mutex initialization failed");
+    rc = pthread_mutex_init(&lf->output_mutex, NULL);
+    if (rc != 0)
+        perror("Mutex initialization failed");
+    lf->active_tasks = 0;
+    lf->shut_down = 0;
+    lf->file_count = 0;
+    lf->error_count = 0;
+    lf->termination_status = TS_ERROR;
+    lf->nthreads = 0;
+    lf->ignore_case = false;
+    lf->sort = false;
+    lf->sort_reverse = false;
+    lf->include_hidden = false; // By default, hidden files are suppressed. Use
@@ -424,3 +433,3 @@
-    f->hidden_only = false;
-    f->flags |= LF_HIDE;     // Turn hide flag on
-    f->follow_links = false; // By default, symbolic links are not followed. Use
+    lf->hidden_only = false;
+    lf->flags |= LF_HIDE;     // Turn hide flag on
+    lf->follow_links = false; // By default, symbolic links are not followed. Use
@@ -428,2 +437,3 @@
-    f->count = false;
-    f->count_silently = false;
+    lf->count = false;
+    lf->count_silently = false;
+
@@ -431 +441 @@
-    argp_parse(&argp, argc, argv, 0, 0, f);
+    argp_parse(&argp, argc, argv, 0, 0, lf);
@@ -436 +446 @@
-            f->base_path = strdup(tmp_str);
+            lf->base_path = strdup(tmp_str);
@@ -438,2 +448,2 @@
-            f->re = strdup(lfargs[0]);
-            f->flags |= LF_REGEX;
+            lf->re = strdup(lfargs[0]);
+            lf->flags |= LF_REGEX;
@@ -445 +455 @@
-            exit(termination_status);
+            exit(lf->termination_status);
@@ -449 +459 @@
-        if (!f->base_path || f->base_path[0] == '\0') {
+        if (!lf->base_path || lf->base_path[0] == '\0') {
@@ -453 +463 @@
-                f->base_path = strdup(tmp_str);
+                lf->base_path = strdup(tmp_str);
@@ -455,3 +465,3 @@
-        if ((!(f->flags & LF_REGEX)) && is_valid_regex(lfargs[1])) {
-            f->re = strdup(lfargs[1]);
-            f->flags |= LF_REGEX;
+        if ((!(lf->flags & LF_REGEX)) && is_valid_regex(lfargs[1])) {
+            lf->re = strdup(lfargs[1]);
+            lf->flags |= LF_REGEX;
@@ -463 +473 @@
-            exit(termination_status);
+            exit(lf->termination_status);
@@ -466,5 +476,5 @@
-    if (f->base_path == nullptr || f->base_path[0] == '\0')
-        f->base_path = strdup(".");
-    termination_status = TS_SUCCESS;
-    if (!f->sort) {
-        init_find(f, argc, argv);
+    if (lf->base_path == nullptr || lf->base_path[0] == '\0')
+        lf->base_path = strdup(".");
+    lf->termination_status = TS_SUCCESS;
+    if (!lf->sort) {
+        init_find(lf, argc, argv);
@@ -472,3 +482,3 @@
-        sort_lf_output(f, argc, argv);
-    if (f->count) {
-        size_t count = atomic_load(&file_count);
+        sort_lf_output(lf, argc, argv);
+    if (lf->count) {
+        size_t count = atomic_load(&lf->file_count);
@@ -477,4 +487,4 @@
-    atomic_load(&error_count);
-    if (error_count > 0) {
-        fprintf(stderr, "Errors: %zu\n", error_count);
-        termination_status |= TS_ERROR;
+    atomic_load(&lf->error_count);
+    if (lf->error_count > 0) {
+        fprintf(stderr, "Errors: %zu\n", lf->error_count);
+        lf->termination_status |= TS_ERROR;
@@ -482 +492 @@
-    exit(termination_status);
+    exit(lf->termination_status);
@@ -492 +502 @@
-void sort_lf_output(SearchFilters *f, int argc, char **argv) {
+void sort_lf_output(LfContext *lf, int argc, char **argv) {
@@ -496 +506 @@
-    if (f->sort_reverse)
+    if (lf->sort_reverse)
@@ -514 +524 @@
-        exit(termination_status);
+        exit(lf->termination_status);
@@ -520 +530 @@
-    init_find(f, argc, argv);            // Initialize and transfer control to the finder
+    init_find(lf, argc, argv);           // Initialize and transfer control to the finder
@@ -554 +564 @@
-bool init_find(SearchFilters *f, int argc, char **argv) {
+bool init_find(LfContext *lf, int argc, char **argv) {
@@ -556,4 +566,4 @@
-    if (!f->include_types)
-        f->include_types = 0xff;
-    if (f->include_types)
-        f->suppress_types = f->include_types ^ 0xff;
+    if (!lf->include_types)
+        lf->include_types = 0xff;
+    if (lf->include_types)
+        lf->suppress_types = lf->include_types ^ 0xff;
@@ -562 +572 @@
-    f->include_hidden = !(f->flags & LF_HIDE);
+    lf->include_hidden = !(lf->flags & LF_HIDE);
@@ -564,5 +574,5 @@
-    f->reg_flags = REG_EXTENDED;
-    if (f->flags & LF_ICASE)
-        f->reg_flags |= REG_ICASE;
-    if (f->flags & LF_REGEX) {
-        reti = regcomp(&f->compiled_re, f->re, f->reg_flags);
+    lf->reg_flags = REG_EXTENDED;
+    if (lf->ignore_case)
+        lf->reg_flags |= REG_ICASE;
+    if (lf->flags & LF_REGEX) {
+        reti = regcomp(&lf->compiled_re, lf->re, lf->reg_flags);
@@ -570,2 +580,2 @@
-            fprintf(stderr, "lf: '%s' Invalid pattern\n", f->re);
-            regfree(&f->compiled_re);
+            fprintf(stderr, "lf: '%s' Invalid pattern\n", lf->re);
+            regfree(&lf->compiled_re);
@@ -575,2 +585,2 @@
-    if (f->flags & LF_EXC_REGEX) {
-        reti = regcomp(&f->compiled_ere, f->ere, f->reg_flags);
+    if (lf->flags & LF_EXC_REGEX) {
+        reti = regcomp(&lf->compiled_ere, lf->ere, lf->reg_flags);
@@ -578,2 +588,2 @@
-            fprintf(stderr, "lf: '%s' Invalid exclude pattern\n", f->ere);
-            regfree(&f->compiled_ere);
+            fprintf(stderr, "lf: '%s' Invalid exclude pattern\n", lf->ere);
+            regfree(&lf->compiled_ere);
@@ -587,2 +597,2 @@
-    if (nthreads == 0) {
-        nthreads = (nprocs > 1) ? (nprocs - 1) : 1;
+    if (lf->nthreads == 0) {
+        lf->nthreads = (nprocs > 1) ? (nprocs - 1) : 1;
@@ -590,4 +600,4 @@
-        if (nthreads < 1)
-            nthreads = 1;
-        if (nthreads > nprocs)
-            nthreads = nprocs;
+        if (lf->nthreads < 1)
+            lf->nthreads = 1;
+        if (lf->nthreads > nprocs)
+            lf->nthreads = nprocs;
@@ -595 +605 @@
-    debug_out(f, argc, argv, nthreads);
+    debug_out(lf, argc, argv);
@@ -598 +608 @@
-    termination_status = TS_SUCCESS;
+    lf->termination_status = TS_SUCCESS;
@@ -601 +611 @@
-    if (stat(f->base_path, &st) == 0) {
+    if (stat(lf->base_path, &st) == 0) {
@@ -604 +614 @@
-            child_task->dir_path = strdup(f->base_path);
+            child_task->dir_path = strdup(lf->base_path);
@@ -610,4 +620,13 @@
-            enqueue_dir(child_task);
-            pthread_t threads[nthreads];
-            for (unsigned int i = 0; i < nthreads; i++) {
-                rc = pthread_create(&threads[i], NULL, finder, f);
+            enqueue_dir(lf, child_task);
+            lf->threads = calloc(lf->nthreads, sizeof(*lf->threads));
+            if (!lf->threads) {
+                fprintf(stderr, "Out of memory allocating threads\n");
+                return false;
+            }
+            for (unsigned int i = 0; i < lf->nthreads; i++) {
+                rc = pthread_create(
+                    &lf->threads[i],
+                    NULL,
+                    finder,
+                    lf);
+
@@ -616,3 +635,3 @@
-                    shut_down = 1;
-                    termination_status = TS_ERROR;
-                    return 0;
+                    lf->shut_down = 1;
+                    lf->termination_status = TS_ERROR;
+                    return false;
@@ -621,7 +640,6 @@
-            pthread_mutex_lock(&queue_mutex);
-            while (!shut_down) {
-                pthread_cond_wait(&cond_var, &queue_mutex);
-            }
-            pthread_mutex_unlock(&queue_mutex);
-            for (unsigned int i = 0; i < nthreads; i++)
-                pthread_join(threads[i], NULL);
+            pthread_mutex_lock(&lf->queue_mutex);
+            while (!lf->shut_down)
+                pthread_cond_wait(&lf->cond_var, &lf->queue_mutex);
+            pthread_mutex_unlock(&lf->queue_mutex);
+            for (unsigned int i = 0; i < lf->nthreads; i++)
+                pthread_join(lf->threads[i], NULL);
@@ -632 +650,3 @@
-                    f->base_path);
+                    lf->base_path);
+            lf->termination_status = TS_ERROR;
+            return false;
@@ -637,11 +657,17 @@
-    if (f->flags & LF_REGEX) {
-        regfree(&f->compiled_re);
-    }
-    if (f->flags & LF_EXC_REGEX) {
-        regfree(&f->compiled_ere);
-    }
-    free(f->base_path);
-    free(f->user_name);
-    free(f->re);
-    free(f->ere);
-    free(f);
+    rc = pthread_mutex_destroy(&lf->queue_mutex);
+    rc = pthread_cond_destroy(&lf->cond_var);
+    rc = pthread_mutex_destroy(&lf->output_mutex);
+    if (rc != 0)
+        perror("Mutex initialization failed");
+    if (lf->flags & LF_REGEX) {
+        regfree(&lf->compiled_re);
+    }
+    if (lf->flags & LF_EXC_REGEX) {
+        regfree(&lf->compiled_ere);
+    }
+    free(lf->base_path);
+    free(lf->user_name);
+    free(lf->re);
+    free(lf->ere);
+    free(lf->threads);
+    free(lf);
@@ -661 +687 @@
-void debug_out(SearchFilters *f, int argc, char **argv, int threads) {
+void debug_out(LfContext *lf, int argc, char **argv) {
@@ -667 +693 @@
-    if (f->debug && (f->report_config || f->report_info || f->report_all)) {
+    if (lf->debug && (lf->report_config || lf->report_info || lf->report_all)) {
@@ -686 +712 @@
-                f->debug ? "true" : "     false");
+                lf->debug ? "true" : "     false");
@@ -688 +714 @@
-                f->report_config ? "true" : "|    false");
+                lf->report_config ? "true" : "|    false");
@@ -690 +716 @@
-                f->report_info ? "true" : "|    false");
+                lf->report_info ? "true" : "|    false");
@@ -692 +718 @@
-                f->report_warnings ? "true" : "|    false");
+                lf->report_warnings ? "true" : "|    false");
@@ -694 +720 @@
-                f->report_errors ? "true" : "|    false");
+                lf->report_errors ? "true" : "|    false");
@@ -696 +722 @@
-                f->report_trace ? "true" : "|    false");
+                lf->report_trace ? "true" : "|    false");
@@ -698 +724 @@
-                f->report_trace ? "true" : "|    false");
+                lf->report_trace ? "true" : "|    false");
@@ -700 +726 @@
-                f->report_all ? "true" : "|    false");
+                lf->report_all ? "true" : "|    false");
@@ -702 +728 @@
-                f->only_errors ? "true" : "|    false");
+                lf->only_errors ? "true" : "|    false");
@@ -704,2 +730,2 @@
-        fprintf(stderr, "Count files: %s\n", f->count ? "true" : "false");
-        fprintf(stderr, "Count only: %s\n", f->count_silently ? "true" : "false");
+        fprintf(stderr, "Count files: %s\n", lf->count ? "true" : "false");
+        fprintf(stderr, "Count only: %s\n", lf->count_silently ? "true" : "false");
@@ -707,2 +733,2 @@
-        fprintf(stderr, "Search directory: %s\n\n", f->base_path);
-        if (f->max_depth == 0)
+        fprintf(stderr, "Search directory: %s\n\n", lf->base_path);
+        if (lf->max_depth == 0)
@@ -711,2 +737,2 @@
-            fprintf(stderr, "Max depth %d\n\n", f->max_depth);
-        fprintf(stderr, "Using %d threads\n\n", threads);
+            fprintf(stderr, "Max depth %d\n\n", lf->max_depth);
+        fprintf(stderr, "Using %d threads\n\n", lf->nthreads);
@@ -715,8 +741,8 @@
-        print_file_type(f->include_types, LF_FIFO, DT_FIFO, "FIFO    p-named pipe");
-        print_file_type(f->include_types, LF_CHR, DT_CHR, "CHR     c-character device");
-        print_file_type(f->include_types, LF_DIR, DT_DIR, "DIR     d-directory");
-        print_file_type(f->include_types, LF_BLK, DT_BLK, "BLK     b-block device");
-        print_file_type(f->include_types, LF_REG, DT_REG, "REG     f-regular file");
-        print_file_type(f->include_types, LF_LNK, DT_LNK, "LINK    l-symbolic link");
-        print_file_type(f->include_types, LF_SOCK, DT_SOCK, "SOCK    s-socket");
-        print_file_type(f->include_types, LF_UNKNOWN, DT_UNKNOWN, "UNKNOWN u-unknown");
+        print_file_type(lf->include_types, LF_FIFO, DT_FIFO, "FIFO    p-named pipe");
+        print_file_type(lf->include_types, LF_CHR, DT_CHR, "CHR     c-character device");
+        print_file_type(lf->include_types, LF_DIR, DT_DIR, "DIR     d-directory");
+        print_file_type(lf->include_types, LF_BLK, DT_BLK, "BLK     b-block device");
+        print_file_type(lf->include_types, LF_REG, DT_REG, "REG     f-regular file");
+        print_file_type(lf->include_types, LF_LNK, DT_LNK, "LINK    l-symbolic link");
+        print_file_type(lf->include_types, LF_SOCK, DT_SOCK, "SOCK    s-socket");
+        print_file_type(lf->include_types, LF_UNKNOWN, DT_UNKNOWN, "UNKNOWN u-unknown");
@@ -724,2 +750,2 @@
-        fprintf(stderr, "f->include_types  = %08b\n", f->include_types);
-        fprintf(stderr, "f->suppress_types = %08b\n", f->suppress_types);
+        fprintf(stderr, "f->include_types  = %08b\n", lf->include_types);
+        fprintf(stderr, "f->suppress_types = %08b\n", lf->suppress_types);
@@ -727,4 +753,4 @@
-        if (f->flags & LF_USER)
-            fprintf(stderr, "User: %s (%ju)\n", f->user_name, f->user_id);
-        if (f->include_perms) {
-            if (f->include_perms & LF_IXUSR)
+        if (lf->flags & LF_USER)
+            fprintf(stderr, "User: %s (%ju)\n", lf->user_name, lf->user_id);
+        if (lf->include_perms) {
+            if (lf->include_perms & LF_IXUSR)
@@ -732 +758 @@
-            if (f->include_perms & LF_IWUSR)
+            if (lf->include_perms & LF_IWUSR)
@@ -734 +760 @@
-            if (f->include_perms & LF_IRUSR)
+            if (lf->include_perms & LF_IRUSR)
@@ -736 +762 @@
-            if (f->include_perms & LF_ISUID)
+            if (lf->include_perms & LF_ISUID)
@@ -738 +764 @@
-            if (f->include_perms & LF_ISGID)
+            if (lf->include_perms & LF_ISGID)
@@ -742,4 +768,4 @@
-        if (f->flags & LF_REGEX)
-            fprintf(stderr, "Include regex: %s\n\n", f->re);
-        if (f->flags & LF_EXC_REGEX)
-            fprintf(stderr, "Exclude regex: %s\n\n", f->ere);
+        if (lf->flags & LF_REGEX)
+            fprintf(stderr, "Include regex: %s\n\n", lf->re);
+        if (lf->flags & LF_EXC_REGEX)
+            fprintf(stderr, "Exclude regex: %s\n\n", lf->ere);
@@ -747 +773 @@
-        if (f->after) {
+        if (lf->after) {
@@ -749 +775 @@
-            format_local_timestamp(f->after, buf, sizeof(buf));
+            format_local_timestamp(lf->after, buf, sizeof(buf));
@@ -753 +779 @@
-        if (f->before) {
+        if (lf->before) {
@@ -755 +781 @@
-            format_local_timestamp(f->before, buf, sizeof(buf));
+            format_local_timestamp(lf->before, buf, sizeof(buf));
@@ -759 +785 @@
-        if (f->file_size_min) {
+        if (lf->file_size_min) {
@@ -761 +787 @@
-            off_t size = f->file_size_min;
+            off_t size = lf->file_size_min;
@@ -771,3 +797,3 @@
-        if (f->max_depth)
-            fprintf(stderr, "Max depth: %d\n\n", f->max_depth);
-        if (f->ignore_case)
+        if (lf->max_depth)
+            fprintf(stderr, "Max depth: %d\n\n", lf->max_depth);
+        if (lf->ignore_case)
@@ -775 +801 @@
-        if (f->include_hidden)
+        if (lf->include_hidden)
@@ -777 +803 @@
-        if (f->follow_links)
+        if (lf->follow_links)
@@ -779 +805 @@
-        if (f->sort)
+        if (lf->sort)
@@ -781 +807 @@
-        if (f->sort_reverse)
+        if (lf->sort_reverse)
@@ -783 +809 @@
-        if (f->report_config && !f->report_all)
+        if (lf->report_config && !lf->report_all)
@@ -798,3 +824,3 @@
-void enqueue_dir(TaskNode *new_task) {
-    pthread_mutex_lock(&queue_mutex);
-    if (qtail)
+void enqueue_dir(LfContext *lf, TaskNode *new_task) {
+    pthread_mutex_lock(&lf->queue_mutex);
+    if (lf->qtail)
@@ -805 +831 @@
-        qtail->next_task = new_task;
+        lf->qtail->next_task = new_task;
@@ -810 +836 @@
-        qhead = new_task;
+        lf->qhead = new_task;
@@ -813 +839 @@
-    qtail = new_task;
+    lf->qtail = new_task;
@@ -819,2 +845,2 @@
-    pthread_cond_signal(&cond_var);
-    pthread_mutex_unlock(&queue_mutex);
+    pthread_cond_signal(&lf->cond_var);
+    pthread_mutex_unlock(&lf->queue_mutex);
@@ -831,2 +857,2 @@
-TaskNode *dequeue_dir() {
-    pthread_mutex_lock(&queue_mutex);
+TaskNode *dequeue_dir(LfContext *lf) {
+    pthread_mutex_lock(&lf->queue_mutex);
@@ -844 +870 @@
-    while (qhead == NULL && !shut_down) {
+    while (lf->qhead == NULL && !lf->shut_down) {
@@ -852,2 +878,2 @@
-        if (atomic_load(&active_tasks) == 0) {
-            shut_down = 1;
+        if (atomic_load(&lf->active_tasks) == 0) {
+            lf->shut_down = 1;
@@ -855 +881 @@
-            pthread_cond_broadcast(&cond_var);
+            pthread_cond_broadcast(&lf->cond_var);
@@ -865 +891 @@
-        pthread_cond_wait(&cond_var, &queue_mutex);
+        pthread_cond_wait(&lf->cond_var, &lf->queue_mutex);
@@ -867,2 +893,2 @@
-    if (shut_down && qhead == NULL) {
-        pthread_mutex_unlock(&queue_mutex);
+    if (lf->shut_down && lf->qhead == NULL) {
+        pthread_mutex_unlock(&lf->queue_mutex);
@@ -871 +897 @@
-    TaskNode *temp = qhead;
+    TaskNode *temp = lf->qhead;
@@ -878,5 +904,5 @@
-    qhead = qhead->next_task;
-    if (!qhead)
-        qtail = NULL;
-    atomic_fetch_add(&active_tasks, 1);
-    pthread_mutex_unlock(&queue_mutex);
+    lf->qhead = lf->qhead->next_task;
+    if (!lf->qhead)
+        lf->qtail = NULL;
+    atomic_fetch_add(&lf->active_tasks, 1);
+    pthread_mutex_unlock(&lf->queue_mutex);
@@ -899 +925 @@
-    SearchFilters *f = (SearchFilters *)arg;
+    LfContext *lf = (LfContext *)arg;
@@ -901 +926,0 @@
-    // regmatch_t pmatch;
@@ -904 +929 @@
-        TaskNode *current_task = dequeue_dir();
+        TaskNode *current_task = dequeue_dir(lf);
@@ -918,0 +944,3 @@
+        // --------------------------------------------------------------------
+        // INITIALIZE DIRECTORY - PRIMING READ
+        // --------------------------------------------------------------------
@@ -922,4 +950,4 @@
-            atomic_fetch_add(&error_count, 1);
-            if (f->debug && (f->report_warnings || f->report_errors ||
-                             f->report_badlinks || f->report_all)) {
-                pthread_mutex_lock(&output_mutex);
+            atomic_fetch_add(&lf->error_count, 1);
+            if (lf->debug && (lf->report_warnings || lf->report_errors ||
+                              lf->report_badlinks || lf->report_all)) {
+                pthread_mutex_lock(&lf->output_mutex);
@@ -927 +955 @@
-                pthread_mutex_unlock(&output_mutex);
+                pthread_mutex_unlock(&lf->output_mutex);
@@ -932,2 +960,2 @@
-            atomic_fetch_sub(&active_tasks, 1);
-            pthread_cond_broadcast(&cond_var);
+            atomic_fetch_sub(&lf->active_tasks, 1);
+            pthread_cond_broadcast(&lf->cond_var);
@@ -938,4 +966,4 @@
-            atomic_fetch_add(&error_count, 1);
-            if (f->debug && (f->report_warnings || f->report_errors ||
-                             f->report_badlinks || f->report_all)) {
-                pthread_mutex_lock(&output_mutex);
+            atomic_fetch_add(&lf->error_count, 1);
+            if (lf->debug && (lf->report_warnings || lf->report_errors ||
+                              lf->report_badlinks || lf->report_all)) {
+                pthread_mutex_lock(&lf->output_mutex);
@@ -944 +972 @@
-                pthread_mutex_unlock(&output_mutex);
+                pthread_mutex_unlock(&lf->output_mutex);
@@ -950,2 +978,2 @@
-            atomic_fetch_sub(&active_tasks, 1);
-            pthread_cond_broadcast(&cond_var);
+            atomic_fetch_sub(&lf->active_tasks, 1);
+            pthread_cond_broadcast(&lf->cond_var);
@@ -954,2 +982,12 @@
-        // unsigned char real_type;
-        //-------------------------------------------------------------
+        //--------------------------------------------------------------------
+        // MAIN LOOP - READ DIRECTORY ENTRIES
+        //--------------------------------------------------------------------
+        // Read the directory entries and process each one. We use readdir to
+        // iterate over the entries in the directory. For each entry, we
+        // construct the full path and use fstatat to get the metadata of
+        // the entry. If the entry is a symbolic link, we check if the user
+        // has chosen to follow links and get the metadata of the target it
+        // points to. We then determine the effective type of the entry and
+        // apply the specified filters (e.g., hidden files, max depth) to
+        // decide whether to process it further or enqueue it for searching.
+        //--------------------------------------------------------------------
@@ -961,2 +999,3 @@
-            stpcpy(stpcpy(stpcpy(full_path, current_task->dir_path), "/"),
-                   entry->d_name);
+            strnz__cpy(full_path, current_task->dir_path, PATH_MAX - 1);
+            strnz__cat(full_path, "/", PATH_MAX - 1);
+            strnz__cat(full_path, entry->d_name, PATH_MAX - 1);
@@ -971,0 +1011 @@
+            effective_type = entry->d_type;
@@ -974,4 +1014,4 @@
-                atomic_fetch_add(&error_count, 1);
-                if (f->debug && (f->report_errors || f->report_warnings ||
-                                 f->report_badlinks || f->report_all)) {
-                    pthread_mutex_lock(&output_mutex);
+                atomic_fetch_add(&lf->error_count, 1);
+                if (lf->debug && (lf->report_errors || lf->report_warnings ||
+                                  lf->report_badlinks || lf->report_all)) {
+                    pthread_mutex_lock(&lf->output_mutex);
@@ -980 +1020 @@
-                    pthread_mutex_unlock(&output_mutex);
+                    pthread_mutex_unlock(&lf->output_mutex);
@@ -984,0 +1025,3 @@
+            if (entry->d_type != effective_type) {
+                fprintf(stderr, "entry->d_type %08b, effective_type %08b\n", entry->d_type, effective_type);
+            }
@@ -996 +1039 @@
-            if (S_ISLNK(st.st_mode)) {
+            // if (effective_type == DT_LNK) {
@@ -997,0 +1041 @@
+            if (S_ISLNK(st.st_mode)) {
@@ -1000,4 +1044,4 @@
-                    atomic_fetch_add(&error_count, 1);
-                    if (f->debug && (f->report_all || f->report_warnings ||
-                                     f->report_errors || f->report_badlinks)) {
-                        pthread_mutex_lock(&output_mutex);
+                    atomic_fetch_add(&lf->error_count, 1);
+                    if (lf->debug && (lf->report_all || lf->report_warnings ||
+                                      lf->report_errors || lf->report_badlinks)) {
+                        pthread_mutex_lock(&lf->output_mutex);
@@ -1006 +1050 @@
-                        pthread_mutex_unlock(&output_mutex);
+                        pthread_mutex_unlock(&lf->output_mutex);
@@ -1010 +1054 @@
-                if (f->follow_links)
+                if (lf->follow_links)
@@ -1015 +1059 @@
-                    if (!f->include_hidden)
+                    if (!lf->include_hidden)
@@ -1017 +1061 @@
-                } else if (f->hidden_only)
+                } else if (lf->hidden_only)
@@ -1038,2 +1082,2 @@
-                if (f->debug && (f->report_trace || f->report_all)) {
-                    pthread_mutex_lock(&output_mutex);
+                if (lf->debug && (lf->report_trace || lf->report_all)) {
+                    pthread_mutex_lock(&lf->output_mutex);
@@ -1041 +1085 @@
-                    pthread_mutex_unlock(&output_mutex);
+                    pthread_mutex_unlock(&lf->output_mutex);
@@ -1044,2 +1088,2 @@
-                    if (f->debug && (f->report_trace || f->report_all)) {
-                        pthread_mutex_lock(&output_mutex);
+                    if (lf->debug && (lf->report_trace || lf->report_all)) {
+                        pthread_mutex_lock(&lf->output_mutex);
@@ -1052 +1096 @@
-                        pthread_mutex_unlock(&output_mutex);
+                        pthread_mutex_unlock(&lf->output_mutex);
@@ -1061,4 +1105,4 @@
-                    atomic_fetch_add(&error_count, 1);
-                    if (f->debug && (f->report_warnings || f->report_errors ||
-                                     f->report_trace || f->report_badlinks ||
-                                     f->report_all)) {
+                    atomic_fetch_add(&lf->error_count, 1);
+                    if (lf->debug && (lf->report_warnings || lf->report_errors ||
+                                      lf->report_trace || lf->report_badlinks ||
+                                      lf->report_all)) {
@@ -1067 +1111 @@
-                        pthread_mutex_lock(&output_mutex);
+                        pthread_mutex_lock(&lf->output_mutex);
@@ -1074 +1118 @@
-                        pthread_mutex_unlock(&output_mutex);
+                        pthread_mutex_unlock(&lf->output_mutex);
@@ -1092 +1136 @@
-                if (f->max_depth != 0 && current_task->depth + 1 == f->max_depth)
+                if (lf->max_depth != 0 && current_task->depth + 1 == lf->max_depth)
@@ -1106 +1150 @@
-                enqueue_dir(child_task);
+                enqueue_dir(lf, child_task);
@@ -1108 +1152 @@
-            if (f->hidden_only && !is_hidden(entry->d_name))
+            if (lf->hidden_only && !is_hidden(entry->d_name))
@@ -1110 +1154 @@
-            scan_file(full_path, f, effective_type);
+            scan_file(full_path, lf, effective_type);
@@ -1116 +1160 @@
-        atomic_fetch_sub(&active_tasks, 1);
+        atomic_fetch_sub(&lf->active_tasks, 1);
@@ -1146 +1190 @@
-int scan_file(char *file_spec, const SearchFilters *f,
+int scan_file(char *file_spec, LfContext *lf,
@@ -1151,2 +1195,2 @@
-        if (f->debug && (f->report_trace || f->report_all)) {
-            pthread_mutex_lock(&output_mutex);
+        if (lf->debug && (lf->report_trace || lf->report_all)) {
+            pthread_mutex_lock(&lf->output_mutex);
@@ -1154,2 +1198,2 @@
-                   f->suppress_types, effective_type, lf_mask[effective_type], f->suppress_types & lf_mask[effective_type], file_spec);
-            pthread_mutex_unlock(&output_mutex);
+                   lf->suppress_types, effective_type, lf_mask[effective_type], lf->suppress_types & lf_mask[effective_type], file_spec);
+            pthread_mutex_unlock(&lf->output_mutex);
@@ -1157 +1201 @@
-        if (f->suppress_types & lf_mask[effective_type])
+        if (lf->suppress_types & lf_mask[effective_type])
@@ -1160 +1204 @@
-        if (f->flags & LF_REGEX) {
+        if (lf->flags & LF_REGEX) {
@@ -1162 +1206 @@
-                regexec(&f->compiled_re, file_spec, 0, NULL, 0);
+                regexec(&lf->compiled_re, file_spec, 0, NULL, 0);
@@ -1165 +1209 @@
-            termination_status |= TS_MATCH;
+            lf->termination_status |= TS_MATCH;
@@ -1168 +1212 @@
-        if (f->flags & LF_EXC_REGEX) {
+        if (lf->flags & LF_EXC_REGEX) {
@@ -1170 +1214 @@
-                regexec(&f->compiled_ere, file_spec, 0, NULL, 0);
+                regexec(&lf->compiled_ere, file_spec, 0, NULL, 0);
@@ -1172 +1216 @@
-                termination_status |= TS_MATCH;
+                lf->termination_status |= TS_MATCH;
@@ -1179 +1223 @@
-        if ((f->flags & LF_USER) && stat(file_spec, &sb) == 0) {
+        if ((lf->flags & LF_USER) && stat(file_spec, &sb) == 0) {
@@ -1181 +1225 @@
-            if (sb.st_uid != f->user_id)
+            if (sb.st_uid != lf->user_id)
@@ -1184 +1228 @@
-        if (f->include_perms) {
+        if (lf->include_perms) {
@@ -1189 +1233 @@
-            if ((f->include_perms & LF_IRUSR) && !(sb.st_mode & S_IRUSR))
+            if ((lf->include_perms & LF_IRUSR) && !(sb.st_mode & S_IRUSR))
@@ -1191 +1235 @@
-            else if ((f->include_perms & LF_IWUSR) && !(sb.st_mode & S_IWUSR))
+            else if ((lf->include_perms & LF_IWUSR) && !(sb.st_mode & S_IWUSR))
@@ -1193 +1237 @@
-            else if ((f->include_perms & LF_IXUSR) && !(sb.st_mode & S_IXUSR))
+            else if ((lf->include_perms & LF_IXUSR) && !(sb.st_mode & S_IXUSR))
@@ -1195 +1239 @@
-            else if ((f->include_perms & LF_ISUID) && !(sb.st_mode & S_ISUID))
+            else if ((lf->include_perms & LF_ISUID) && !(sb.st_mode & S_ISUID))
@@ -1197 +1241 @@
-            else if ((f->include_perms & LF_ISGID) && !(sb.st_mode & S_ISGID))
+            else if ((lf->include_perms & LF_ISGID) && !(sb.st_mode & S_ISGID))
@@ -1200 +1244 @@
-        if (f->before) {
+        if (lf->before) {
@@ -1205 +1249 @@
-            if (stat_cached && sb.st_mtime > f->before)
+            if (stat_cached && sb.st_mtime > lf->before)
@@ -1208 +1252 @@
-        if (f->after) {
+        if (lf->after) {
@@ -1213 +1257 @@
-            if (stat_cached && sb.st_mtime < f->after)
+            if (stat_cached && sb.st_mtime < lf->after)
@@ -1216 +1260 @@
-        if (f->file_size_min) {
+        if (lf->file_size_min) {
@@ -1221 +1265 @@
-            if (stat_cached && sb.st_size < f->file_size_min)
+            if (stat_cached && sb.st_size < lf->file_size_min)
@@ -1224 +1268 @@
-        if (f->only_errors)
+        if (lf->only_errors)
@@ -1226,2 +1270,2 @@
-        atomic_fetch_add(&file_count, 1);
-        if (!f->count_silently) {
+        atomic_fetch_add(&lf->file_count, 1);
+        if (!lf->count_silently) {
@@ -1234 +1278,3 @@
-            pthread_mutex_lock(&output_mutex);
+            outbuf[l] = '\n';
+            outbuf[l + 1] = '\0';
+            pthread_mutex_lock(&lf->output_mutex);
@@ -1236 +1282 @@
-                puts(&outbuf[2]);
+                fputs_unlocked(&outbuf[2], stdout);
@@ -1238,2 +1284,2 @@
-                puts(outbuf);
-            pthread_mutex_unlock(&output_mutex);
+                fputs_unlocked(outbuf, stdout);
+            pthread_mutex_unlock(&lf->output_mutex);
